lib/util/api.ts

```diff
@@ -14,190 +14,280 @@
-ipc.config.stopRetrying = 0;
+// TODO: use the path as cheap authentication
 
-const DISCONNECT_DELAY = 100;
-const IPC_SERVER_ID = process.env.IPC_SERVER_ID as string;
+const wss = new WebSocketServer({ host, port });
 
-/**
- * @summary Send a message to the IPC server
- */
-function emit(channel: string, message?: any) {
-	ipc.of[IPC_SERVER_ID].emit(channel, message);
-}
-
-/**
- * @summary Send a log debug message to the IPC server
- */
-function log(message: string) {
-	if (console?.log) {
-		console.log(message);
-	}
-	emit('log', message);
-}
+// hold emit functions
+let emitLog: (message: string) => void | undefined;
+let emitState: (state: MultiDestinationProgress) => void | undefined;
+let emitFail: (data: any) => void | undefined;
+let emitDrives: (drives: Dictionary<DrivelistDrive>) => void | undefined;
+let emitSourceMetadata: (
+	sourceMetadata: SourceMetadata | Record<string, never>,
+) => void | undefined; // Record<string, never> means an empty object
 
 /**
  * @summary Terminate the child process
  */
-async function terminate(exitCode: number) {
-	ipc.disconnect(IPC_SERVER_ID);
+async function terminate(exitCode?: number) {
 	await cleanup(Date.now());
 	process.nextTick(() => {
 		process.exit(exitCode || SUCCESS);
 	});
 }
 
-/**
- * @summary Handle errors
- */
-async function handleError(error: Error) {
-	emit('error', toJSON(error));
-	await delay(DISCONNECT_DELAY);
-	await terminate(GENERAL_ERROR);
+// kill the process if no initila connections or heartbeat for X sec (default 10)
+function setTerminateTimeout() {
+	if (ETCHER_TERMINATE_TIMEOUT > 0) {
+		return setTimeout(() => {
+			console.log(
+				`no connections or heartbeat for ${ETCHER_TERMINATE_TIMEOUT} ms, terminating`,
+			);
+			terminate();
+		}, ETCHER_TERMINATE_TIMEOUT);
+	} else {
+		return null;
+	}
 }
 
-/**
- * @summary Abort handler
- * @example
- */
-const onAbort = async (exitCode: number) => {
-	log('Abort');
-	emit('abort');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
-};
-
-const onSkip = async (exitCode: number) => {
-	log('Skip validation');
-	emit('skip');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
-};
-
-ipc.connectTo(IPC_SERVER_ID, () => {
-	// Gracefully exit on the following cases. If the parent
-	// process detects that child exit successfully but
-	// no flashing information is available, then it will
-	// assume that the child died halfway through.
+// terminate the process cleanly on SIGINT
+process.once('SIGINT', async () => {
+	await terminate(SUCCESS);
+});
 
-	process.once('uncaughtException', handleError);
+// terminate the process cleanly on SIGTERM
+process.once('SIGTERM', async () => {
+	await terminate(SUCCESS);
+});
 
-	process.once('SIGINT', async () => {
-		await terminate(SUCCESS);
-	});
+let TerminateInterval = setTerminateTimeout();
 
-	process.once('SIGTERM', async () => {
-		await terminate(SUCCESS);
-	});
+type EmitLog = {
```

Comment:
I don’t know whether one is better than the other, but why declare a type alias instead of an interface which IME is the usual choice?

Reply:
👍